### ISSUE TITLE
Reducing docker image size

### ISSUE THREAD

**Issue by Brandl** [tomav/docker-mailserver#98]
Hello,

there are not many solutions for a docker based mailserver out there and from what I've looked at this is a very promising project. Starting at feature-completness to active maintenance this features a lot of the things I'm looking for to use in production.

One thing, that would totally win me over is to reduce the docker image size to a minimum. Currently the image size is around 500mb, which is okay, but could be improved a lot:
https://imagelayers.io/?images=tvial%2Fdocker-mailserver:latest

Here is a checklist of things that worked well:
- [ ] move to a minimal base image like docker alpine (100-200mb reduction)
- [ ] do installs and cleanup in a single RUN-command 
- [ ] reduce number of installed packages (do you really need vim in a container?)

This might include some compatibility breaking changes and requires some fine tuning, like looking for equivalent packages in the docker alpine package index. On the plus side, this would:

- Decrease download size and increase download/build speed
- Increase start/general performance
- Decrease resource usage and footprint
- Improve security by minimising the attack surface

The docker fluentd image, is a excellent showcase, where this worked well:
https://github.com/fluent/fluentd-docker-image/pull/9

**Comment by tomav** [comment]
Hi @Brandl and thanks for your support.
These tasks can be done very quickly:
* do installs and cleanup in a single RUN-command
* reduce number of installed packages (do you really need vim in a container?)

If you wanna PR this, it can be merged very fast.

Regarding the base image, why not! I'll have to have a look to these images. I mostly work with Ubuntu and Debian so I don't know the other.

**Comment by Gymnae** [comment]
I just replied to a closed request going in the same direction.
It would be superb to see this project with an Alpine base and it's smaller attack surface.

**Comment by PhilippeChepy** [comment]
I have removed vim, and merged lots of commands in Dockerfil in #102

**Comment by tomav** [comment]
The move to Alpine will break a lot of changed. 
We can plan that later in a specific branch to test things before merge.
Anyone can work on that?

The rest is being discussed in #102 with @PhilippeChepy.

**Comment by tomav** [comment]
An image can be tested at `tvial/docker-mailserver:lightweight `

**Comment by tomav** [comment]
tvial/docker-mailserver           lightweight                fe2efbb8c8e8        47 hours ago        432.4 MB
    tvial/docker-mailserver           latest                     294dcb629180        2 weeks ago         540 MB

**Comment by tomav** [comment]
Anyone tried?

**Comment by jmparra** [comment]
i have tested with debian, but no improvements in image size
```
REPOSITORY                TAG                 IMAGE ID            CREATED             SIZE
tvial/docker-mailserver   debian              b2e55093fdc4        2 minutes ago       432.5 MB
tvial/docker-mailserver   lightweight         fe2efbb8c8e8        2 days ago          432.4 MB
tvial/docker-mailserver   latest              f8595bb70bb2        3 days ago          540.9 MB
```

**Comment by tomav** [comment]
As proposed above, we can work on a POC with alpine if we think that it will really reduce the image size (based on `docker images`, it's 180MB)
This could be the base of a `docker-mailserver:v2` with some improvements that I have in my head but I will need contributors.

**Comment by tomav** [comment]
Let me know if few of you are ready to help. Thanks.

**Comment by Gymnae** [comment]
I would like to help, just that I'm not a developer. I could point a domain which I don't use for testing to an install. Are there specifics you'd like to be tested? Or how could one assist in moving it to an alpine based image? I've set up some docker images on alpine - I just started with docker some weeks ago - and it feels a bit gentoo'ish - which is fine by me, but can be anal compared to debian based distros.

**Comment by tomav** [comment]
@Gymnae we also need testers :-)

**Comment by KyleOndy** [comment]
@tomav I'd be willing to help with the POC / testing of an alpine based container.

**Comment by tomav** [comment]
Check this: https://github.com/tomav/docker-mailserver/issues/109

**Comment by KyleOndy** [comment]
So I've started to work on the alpine container. A thing that I noticed is that the [opendkim](https://pkgs.alpinelinux.org/package/testing/x86_64/opendkim) and [courier-authlib](https://pkgs.alpinelinux.org/package/testing/x86_64/courier-authlib) packages are both in the testing repository of alpine linux.

I'm not sure if we using packages from a testing repo is a great idea.

**Comment by tomav** [comment]
The time we build this new image, perhaps the packages will be stable. We need some time to refactor some things and test it before out stable release.
But good point and open question.

**Comment by gegere** [comment]
I have a good amount of knowledge with shippable.com. I see the site tests the build using Travis-CI but using Shippable the entire build process could be tested. Thoughts?

**Comment by tomav** [comment]
@gegere what do you mean by "the entire build process could be tested"?

**Comment by tomav** [comment]
@KyleOndy it seems that some other packages are not yet available (even in testing). You confirm?

**Comment by KyleOndy** [comment]
@tomav Confirmed. I am having a really tough time getting anywhere with the alpine container.

**Comment by gegere** [comment]
Regarding my https://github.com/tomav/docker-mailserver/issues/98#issuecomment-203830743 you are using Travis-CI to build and check various merges on the build. Correct? As more and more variables are added various builds would need a matrix of environmental variables to test all the various options. This could be part of the .travis.yml file.

Here is another Docker container and I am using CircleCi and Shippable to manage testing of the builds and verification that everything is working as it should.

Here is a sample of the .yml file I use for this build process.
https://github.com/htmlgraphic/Apache/blob/master/shippable.yml

I understand the need for testing and would very much would like to lend my skills and knowledge.

**Comment by tomav** [comment]
You're right, the way we do it today is not pretty clean.
If you have an idea to test env vars configurations using travis (before trying to move to another CI service), feel free to PR something. I'm gonna add this to v2 requirements.

**Comment by tomav** [comment]
#102 if a first step in size reduction but has ti be finished.

@KyleOndy is there another base image we can use instead of alpine?

**Comment by gegere** [comment]
I am not super familiar with Travis-CI and regarding Docker builds they were late to the game when I was investigating a year ago, this looks to have changed. Recently Shippable released a big update the beginning of March. Many improvements to their backend service and processes which required a large amount of time to refactor my tests on a similar repo and in doing so I realized new methods and functions that really make it the best CI system available.

I have been continuing to compare Circle-CI and Shippable directly as any pull request would allow a build to **spin up within both services**. I have had one CI service fail because of a connection issue to a package download while the other build service continued successfully. 

Basically my point is this. Using docker-compose and other methods should give you the flexibility to test the build with any CI service. As each CI service becomes more robust you can test us minimal effort because commonalty between each service will be very similar.

$ **docker-compose -f docker-compose.whatever.yml up -d**

**Comment by tomav** [comment]
Closing this issue because:

- master has been reduced
- same for v2 on which efforts are concentrated
- an attempt on moving to alpine is available here: https://github.com/tomav/docker-mailserver/pull/151
